lib/E2E/LowerShapeConstraints.cpp

```diff
@@ -0,0 +1,185 @@
+    // lowering e.g. the shape.broadcast op itself" (which technically puts the
+    // errors in some random location in the IR that is not the
+    // shape.to_extent_tensor op). E.g. the following code would miscompile with
+    // either of those ways that these ops get lowered in practice:
+    // ```
+    // %shape = shape.broadcast %lhs, %rhs : !shape.shape
+    // if %cond:
+    //     shape.to_extent_tensor(%shape)
+    // ```
+    // It's not possible to correctly compile this code without significant
+    // contortions (such as carrying an "is error" bit). And to boot, we
+    // shouldn't be getting into that situation in the first place! But the
+    // `shape.to_extent_tensor : !shape.shape -> tensor<?xindex>` abstraction
+    // opens up that possibility.
+    //
+    // shape.to_extent_tensor should not really be a thing, since it creates
+    // these ill-defined situations about where errors are observed. A
+    // !shape.shape type should only exist (for this compilation flow) as part
+    // of a utility, something like "I want to do this shape calculation on
+    // !shape.shape type, create IR that uses tensor<?xindex> and witnesses to
+    // implement it, on the assumption that the error can be
+    // observed anywhere inside the shape calculation".
+    //
+    // !shape.shape type would still be useful for lowerings that actually
+    // result in a runtime type that carries an "is error" bit inside it, though
+    // TBD if such use cases arise.
+    if (op.getType().isa<shape::ShapeType>() ||
+        op.lhs().getType().isa<shape::ShapeType>() ||
+        op.rhs().getType().isa<shape::ShapeType>()) {
+      return op.emitError() << "Error shapes should not exist at this point";
+    }
+
+    auto loc = op.getLoc();
+    Value zero = rewriter.create<ConstantIndexOp>(loc, 0);
+    Value one = rewriter.create<ConstantIndexOp>(loc, 1);
+
+    // Find smaller and greater rank and extent tensor.
+    Value lhsRank = rewriter.create<DimOp>(loc, op.lhs(), zero);
+    Value rhsRank = rewriter.create<DimOp>(loc, op.rhs(), zero);
+    Value lhsSmaller =
+        rewriter.create<CmpIOp>(loc, CmpIPredicate::ule, lhsRank, rhsRank);
+    Type indexTy = rewriter.getIndexType();
+    Type extentTensorTy = op.lhs().getType();
+    auto ifOp = rewriter.create<scf::IfOp>(
+        loc, TypeRange{indexTy, extentTensorTy, indexTy, extentTensorTy},
+        lhsSmaller,
+        [&](OpBuilder &b, Location loc) {
+          b.create<scf::YieldOp>(
+              loc, ValueRange{lhsRank, op.lhs(), rhsRank, op.rhs()});
+        },
+        [&](OpBuilder &b, Location loc) {
+          b.create<scf::YieldOp>(
+              loc, ValueRange{rhsRank, op.rhs(), lhsRank, op.lhs()});
+        });
+    Value smallerRank = ifOp.getResult(0);
+    Value smallerOperand = ifOp.getResult(1);
+    Value greaterRank = ifOp.getResult(2);
+    Value greaterOperand = ifOp.getResult(3);
+
+    Value rankDiff =
```

Comment:
> ```
>     // Compare the shapes extent by extent, and emit errors for
>     // non-broadcast-compatible shapes.
> ```
I read the open-coded logic below for quite a while to verify it conforms with the formula as I know it. it might be useful to actually write that out as a comment:

"Two extents are broadcastValid if a) they are both equal, or b) at least one of them is 1."

Also it confused me for a bit that "greater" and "smaller" refer to the greater-ranked extents and smaller-ranked extents, not the greater or smaller extent.

Reply:
Agreed. Very confusing on a re-read. Updated to somewhat more verbose "greaterRankOperand"/"lesserRankOperand".